vpr/src/place/move_utils.cpp

```diff
@@ -664,3 +665,326 @@ bool placer_breakpoint_reached() {
+        //bound the range of valid blocks at this x-location, which
+        //are within y range
+        std::unordered_set<int> tried_dy;
+        while (!legal && (int)tried_dy.size() < y_range) { //Until legal or all possibilities exhausted
+            //Randomly pick a y location
+            int dy = vtr::irand(y_range - 1);
+
+            //Record this y location as tried
+            auto res2 = tried_dy.insert(dy);
+            if (!res2.second) {
+                continue; //Already tried this position
+            }
+
+            //Key in the y-dimension is the compressed index location
+            cy_to = (y_lower_iter + dy)->first;
+
+            VTR_ASSERT(cy_to >= min_cy);
+            VTR_ASSERT(cy_to <= max_cy);
+
+            if (cx_from == cx_to && cy_from == cy_to) {
+                continue; //Same from/to location -- try again for new y-position
+            } else {
+                legal = true;
+            }
+        }
+    }
+
+    if (!legal) {
+        //No valid position found
+        return false;
+    }
+
+    VTR_ASSERT(cx_to != OPEN);
+    VTR_ASSERT(cy_to != OPEN);
+
+    //Convert to true (uncompressed) grid locations
+    to.x = compressed_block_grid.compressed_to_grid_x[cx_to];
+    to.y = compressed_block_grid.compressed_to_grid_y[cy_to];
+
+    auto& grid = g_vpr_ctx.device().grid;
+
+    auto to_type = grid[to.x][to.y].type;
+
+    //Each x/y location contains only a single type, so we can pick a random
+    //z (capcity) location
+    auto& compatible_sub_tiles = compressed_block_grid.compatible_sub_tiles_for_tile.at(to_type->index);
+    to.sub_tile = compatible_sub_tiles[vtr::irand((int)compatible_sub_tiles.size() - 1)];
+
+    VTR_ASSERT_MSG(is_tile_compatible(to_type, type), "Type must be compatible");
+    VTR_ASSERT_MSG(grid[to.x][to.y].width_offset == 0, "Should be at block base location");
+    VTR_ASSERT_MSG(grid[to.x][to.y].height_offset == 0, "Should be at block base location");
+
+    return true;
+}
+
+bool find_to_loc_centroid(t_logical_block_type_ptr type,
+                          float rlim,
+                          const t_pl_loc from,
+                          const t_pl_loc centroid,
+                          t_pl_loc& to,
+                          int dm_rlim) {
+    //Finds a legal swap to location for the given type, starting from 'from.x' and 'from.y'
+    //
+    //Note that the range limit (rlim) is applied in a logical sense (i.e. 'compressed' grid space consisting
+    //of the same block types, and not the physical grid space). This means, for example, that columns of 'rare'
+    //blocks (e.g. DSPs/RAMs) which are physically far appart but logically adjacent will be swappable even
+    //at an rlim fo 1.
+    //
+    //This ensures that such blocks don't get locked down too early during placement (as would be the
+    //case with a physical distance rlim)
+
+    //catch the initial rlim
+    static float init_rlim = -1;
```

Comment:
I think it would be better to make this a global at the top of the file, so it's easier to see and document.

Reply:
Agree!